File: app/src/main/java/kr/djspi/pipe01/fragment/MessageDialog.java
package kr.djspi.pipe01.fragment;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.TextView;

import kr.djspi.pipe01.MainActivity;
import kr.djspi.pipe01.R;

import static android.content.Intent.CATEGORY_DEFAULT;
import static android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS;
import static android.provider.Settings.ACTION_NFC_SETTINGS;
import static android.text.Html.fromHtml;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static java.util.Objects.requireNonNull;

public class MessageDialog extends DialogFragment implements OnClickListener {

    private boolean isReturnToMain = false;
    private int issueType = 0;
    private TextView popupTitle;
    private TextView popupText;
    private TextView popupSubText;
    private TextView buttonDismiss;
    private TextView buttonOk;

    public MessageDialog() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        isReturnToMain = false;
        if (getArguments() != null) {
            issueType = getArguments().getInt("issueType");
        }
    }

    /**
     * 레이아웃 구성 및 기능 초기화. 상황에 따라 팝업 메시지 전환
     */
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_message, container, false);

        popupTitle = view.findViewById(R.id.popup_title);
        popupTitle.setText("알림");
        popupText = view.findViewById(R.id.popup_contents);
        popupSubText = view.findViewById(R.id.popup_contents_sub);
        buttonDismiss = view.findViewById(R.id.btn_dismiss);
        buttonDismiss.setOnClickListener(this); // 취소버튼은 기본적으로 비표시 (invisible) 상태
        buttonOk = view.findViewById(R.id.btn_ok);
        buttonOk.setOnClickListener(v -> dismiss());
        view.findViewById(R.id.btn_close).setOnClickListener(this);

        setPopup(issueType, view);

        return view;
    }

    /**
     * 상황에 따라 알맞는 팝업 메시지를 설정
     *
     * @param issueType 팝업을 띄우는 클래스에 전달하는 상황
     * @see MessageDialog#setVisibilityToGone(View) 보충 설명(Sub) 을 표시하지 않는다.
     * @see #getTag 팝업을 띄우는 클래스에서 전달하는 메시지를 태그로 받아온다.
     */
    private void setPopup(final int issueType, @NonNull final View view) {
        // 팝업창 제목과 내용, 확인 버튼에 Default 값을 먼저 설정
        popupTitle.setText("알림");
        popupText.setText(getTag());
        // 전달된 상황에 맞게 팝업창 커스터마이즈
        switch (issueType) {
            case 0: // 일반 메시지 전달
                setVisibilityToGone(view);
                break;
            case 1: // (공통) 위치 기능이 꺼져 있음
                popupTitle.setText("주의");
                popupSubText.setText(fromHtml(getString(R.string.popup_location_on_sub)));
                // 확인버튼 터치 시 위치 설정으로 이동
                buttonOk.setOnClickListener(v -> {
                    startActivity(new Intent(ACTION_LOCATION_SOURCE_SETTINGS).addCategory(CATEGORY_DEFAULT));
                    dismiss();
                });
                break;
            case 2: // (공통) NFC 기능이 꺼져 있음
                popupTitle.setText("주의");
                popupSubText.setText(fromHtml(getString(R.string.popup_nfc_on_sub)));
                // 확인버튼 터치 시 NFC 설정으로 이동
                buttonOk.setOnClickListener(v -> {
                    startActivity(new Intent(ACTION_NFC_SETTINGS).addCategory(CATEGORY_DEFAULT));
                    dismiss();
                });
                break;
            case 3: // (MainActivity.class) 정품 SPI 가 아닌 태그가 태깅되었음
                setVisibilityToGone(view);
                popupTitle.setText("주의");
                buttonOk.setText("종료");
                buttonOk.setOnClickListener(v -> {
                    dismiss();
                    android.os.Process.killProcess(android.os.Process.myPid());
                    System.exit(1); // 앱 완전종료
                });
                break;
            case 4: // (SpiPostActivity.class) 쓰기 작업 이후 수정이 불가함을 안내
                setVisibilityToGone(view);
                popupTitle.setText("주의");
                buttonDismiss.setVisibility(VISIBLE);
                buttonDismiss.setText(getString(R.string.popup_pre));
                break;
            case 5: // (SpiPostActivity.class) 정보가 정상적으로 기록됨
                setVisibilityToGone(view);
                isReturnToMain = true;
                break;
            case 6: // (SpiPostActivity.class) 하나 이상의 관로 정보 등록 과정에서 에러 발생 안내
                popupText.setText(getString(R.string.popup_error_set));
                popupSubText.setText(getTag());
                break;
            case 7: // (공통) 서버와의 통신 에러
                popupText.setText(getString(R.string.popup_error_comm));
//                popupSubText.setText(getTag());
                setVisibilityToGone(view);
            default:
                break;
        }
    }

    private static void setVisibilityToGone(@NonNull final View view) {
        view.findViewById(R.id.popup_contents_sub).setVisibility(GONE);
        view.findViewById(R.id.icon_info_s).setVisibility(GONE);
        view.findViewById(R.id.view_border).setVisibility(GONE);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_dismiss:
                requireNonNull(getActivity()).onBackPressed();
                break;
            case R.id.btn_close:
                isReturnToMain = true;
                dismiss();
            default:
                break;
        }
    }

    @Override
    public void onDismiss(DialogInterface dialogInterface) {
        if (isReturnToMain) {
            Intent intent = new Intent(getContext(), MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
            if (isAdded()) startActivity(intent);
            else return;
        }
        super.onDismiss(dialogInterface);
    }
}

File: app/src/main/java/kr/djspi/pipe01/fragment/MessageDialog.kt
package kr.djspi.pipe01.fragment

import android.content.DialogInterface
import android.content.Intent
import android.content.Intent.*
import android.os.Bundle
import android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS
import android.provider.Settings.ACTION_NFC_SETTINGS
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import kr.djspi.pipe01.BaseActivity.Companion.isReadyForPost
import kr.djspi.pipe01.MainActivity
import kr.djspi.pipe01.R
import kr.djspi.pipe01.util.fromHtml
import kotlin.system.exitProcess

class MessageDialog : DialogFragment(), OnClickListener {

    private var returnToMain: Boolean = false
    private var issue: Int? = 0
    private lateinit var title: TextView
    private lateinit var contents: TextView
    private lateinit var contentsSub: TextView
    private lateinit var buttonDismiss: TextView
    private lateinit var buttonOk: TextView
    private lateinit var infoIcon: ImageView
    private lateinit var border: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        issue = arguments?.getInt("issueType")
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_message, container, false)
        title = view.findViewById(R.id.popup_title)
        contents = view.findViewById(R.id.popup_contents)
        contentsSub = view.findViewById(R.id.popup_contents_sub)
        buttonDismiss = view.findViewById(R.id.button_dismiss)
        buttonDismiss.setOnClickListener(this)
        buttonOk = view.findViewById(R.id.button_ok)
        buttonOk.setOnClickListener(this)
        infoIcon = view.findViewById(R.id.icon_info_s)
        border = view.findViewById(R.id.view_border)
        view.findViewById<ImageView>(R.id.button_close).setOnClickListener(this)
        setDialog(issue)
        return view
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            R.id.button_dismiss -> activity?.onBackPressed()
            R.id.button_ok -> {
                if (issue == 5) {
                    isReadyForPost = true
                }
                dismiss()
            }
            R.id.button_close -> {
                returnToMain = true
                dismiss()
            }
        }
    }

    private fun setDialog(issue: Int?) {
        title.text = "알림"
        contents.text = tag
        when (issue) {
            0 -> { // 일반 메시지 전달
                setVisibilityToGone()
            }
            1 -> { // (공통) 위치 기능이 꺼져 있음
                title.text = "주의"
                contentsSub.text = fromHtml(getString(R.string.popup_location_on_sub))
                buttonOk.setOnClickListener {
                    startActivity(
                        Intent(ACTION_LOCATION_SOURCE_SETTINGS).addCategory(CATEGORY_DEFAULT)
                    )
                    dismiss()
                }
            }
            2 -> { // (공통) NFC 기능이 꺼져 있음
                title.text = "주의"
                contentsSub.text = fromHtml(getString(R.string.popup_nfc_on_sub))
                buttonOk.setOnClickListener {
                    startActivity(Intent(ACTION_NFC_SETTINGS).addCategory(CATEGORY_DEFAULT))
                    dismiss()
                }
            }
            3 -> { // (MainActivity.kt) 정품 SPI 가 아닌 태그가 태깅되었음
                setVisibilityToGone()
                title.text = "주의"
                buttonOk.text = "종료"
                buttonOk.setOnClickListener {
                    dismiss()
//                    android.os.Process.killProcess(android.os.Process.myPid())
                    System.runFinalization()
                    exitProcess(0)
                }
            }
            4 -> { // (MainActivity.kt) 태그 정보 조회 실패
                setVisibilityToGone()
            }
            5 -> { // (SpiPostActivity.kt) 쓰기 작업 이후 수정이 불가함을 안내
                setVisibilityToGone()
                title.text = "주의"
                contents.text = fromHtml(getString(R.string.popup_read_only))
                buttonDismiss.visibility = View.VISIBLE
                buttonDismiss.text = "이전"
            }
            6 -> { // (SpiPostActivity.kt) 정보가 정상적으로 기록됨
                setVisibilityToGone()
                returnToMain = true
            }
            7 -> { // (SpiPostActivity.kt) 하나 이상의 관로 정보 등록 과정에서 에러 발생 안내
                contents.text = getString(R.string.popup_error_set)
                contentsSub.text = tag
            }
            8 -> { // (공통) 서버와의 통신 에러
                contents.text = getString(R.string.popup_error_comm)
                setVisibilityToGone()
            }
            9 -> { // (MainActivity.kt) 앱 시작 시 절전모드가 실행중인지 확인
                title.text = "주의"
                contentsSub.text = getString(R.string.popup_power_save_sub)
            }
            10 -> { // (BaseActivity.kt) 위치 정보를 가져오지 못하여 지도보기를 실행하지 못함
                title.text = "주의"
                contentsSub.text = getString(R.string.popup_error_location_count_exceed_sub)
            }
            11 -> { // (MainActivity.kt) SPI 제품 초기화 개편 (선로종류, 형태, 관리기관 추가) 후 이전 초기화 제품에 대한 안내
                contents.text = fromHtml(tag!!)
                setVisibilityToGone()
            }
            12 -> { // (SpiPostActivity.kt) 정보를 등록하려는 태그와 실제 태깅된 태그가 다를 때 등록 거부
                contentsSub.text = getString(R.string.popup_error_serial_mismatch_sub)
            }
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        if (returnToMain) {
            if (isAdded) {
                startActivity(
                    Intent(context, MainActivity::class.java)
                        .addFlags(FLAG_ACTIVITY_CLEAR_TOP)
                        .addFlags(FLAG_ACTIVITY_NEW_TASK)
                )
            } else return
        } else {
            return
        }
        isReadyForPost = false
        super.onDismiss(dialog)
    }

    private fun setVisibilityToGone() {
        contentsSub.visibility = View.GONE
        infoIcon.visibility = View.GONE
        border.visibility = View.GONE
    }

    companion object {
        @JvmStatic
        fun getInstance(issue: Int, cancelable: Boolean): MessageDialog {
            return MessageDialog().apply {
                arguments = Bundle().apply {
                    putInt("issueType", issue)
                }
                isCancelable = cancelable
            }
        }
    }
}
